Add Parser skeleton.

fernape opened the pull request:
Also, add definition for the `Part` interface to be implemented by the parts that will make the AST.

The parser contains two indexes to keep track of the current position and the last parsed position.

juanmanuel-tirado reviewed (approved):
R2GO

I think this is a reasonable starting point.

jameinel commented on parse/lexer.go:
```diff
@@ -0,0 +1,212 @@
+// reads it as tokens for consumption by our parser.
+// It operates in a lazy fashion, requiring a call to
+// `NextToken` to advance through the input.
+// Offsets and lines are zero-indexed.
```
almost all text editors use 1 based counting (going to the first line is 1,1 not 0,0).
So any error messages, etc we generate should certainly line up with your most likely source of interpreting the content.
Is the intent for 0,0 so that you apply that as an offset to something like https://pkg.go.dev/runtime#Func.FileLine ?

manadart replied:
No rationale other than uniformity. Simon's first incarnation had 1-based line counting. It also included 1-based columns as well, but I left that out of this initial cut. It may be restored if useful.

manadart replied:
Restored 1-based lines and columns.

fernape commented on parse/token.go:
```diff
@@ -0,0 +1,93 @@
+	SEPARATOR
+
+	IDENT
+	INT //int literal
```
The lexer tokenizes well both 10000 and [10000.5](https://github.com/canonical/sqlair/blob/4dd1d0aa38105a2e289555f77e9843a3697e0849/parse/lexer_test.go#L35).
Maybe we could change INT by NUM.

manadart commented on parse/lexer.go:
```diff
@@ -0,0 +1,212 @@
+				ret = append(ret, l.char)
+			} else {
+				ret = append(ret, l.char)
+
```
Done. Thanks.

benhoyt reviewed (commented):
Nice start! I've added a bunch of code review comments, mainly based on my experience with GoAWK's lexer/parser.

However, stepping back a bit, do we actually need full lexing and parsing here, or could we just look for our special `{...}` and `:name` sequences? I could be wrong, but I think `{` and `}` and `:` can only occur in SQL string literals, right? If so, all we'd need is an "in string" / "not in string" state, and if we're not in a string we look for `{` and `:`. That would avoid us having to do full SQL tokenization, which I suspect is a good bit more complex and probably easy to get wrong.

If we do need a full lexer/parser, would it be worthwhile to reuse an existing one, at least as a base to work from? I think this is fairly complex and easy to get edge cases wrong, so it might be better to use a solid existing parser as a starting point, even if it means copying it across and hacking on it.

Some existing Go ones at a quick search (I haven't evaluated any of these for quality):

* https://github.com/auxten/postgresql-parser/blob/main/pkg/sql/parser/scan.go
* https://github.com/krasun/gosqlparser/blob/main/lexer.go
* https://github.com/blastrain/vitess-sqlparser/blob/develop/sqlparser/token.go
* https://github.com/pingcap/parser/blob/master/lexer.go
* https://github.com/akito0107/xsqlparser/blob/master/sqltoken/tokenizer.go

manadart commented on parse/token.go:
```diff
@@ -0,0 +1,93 @@
+// maybeRuneToken returns a token for the input rune if
+// it from is the subset that we specifically recognise.
+// If unrecognised, zero-value, false is returned.
+func maybeRuneToken(char rune, pos Position) (Token, bool) {
```
Done.

manadart commented on parse/lexer.go:
```diff
@@ -0,0 +1,213 @@
+	return l.input[l.readOffset+n]
+}
+
+func isLetter(char rune) bool {
```
Inlined `unicode.IsLetter`

manadart commented on parse/lexer.go:
```diff
@@ -0,0 +1,213 @@
+func (l *Lexer) readIdentifier() string {
+	pos := l.offset
+
+	for isLetter(l.char) || isDigit(l.char) || l.char == '-' {
```
`-` is actually invalid in a SQL identifier.

manadart commented on parse/lexer.go:
```diff
@@ -0,0 +1,213 @@
+// NewLexer creates a new Lexer from a given input and primes it
+// with the first non-whitespace character before returning.
+func NewLexer(input string) *Lexer {
+	l := &Lexer{input: []rune(strings.TrimSpace(input))}
```
There's give and take on cogency when adding `l.char, _ = utf8.DecodeRuneInString(string(l.input[l.readOffset]))`...

I'm OK with mutating the original statement with the initial trim. It's more to allow the engineer to construct visually coherent statements when set against `gofmt`. Example from test:
```
	stmt := `
SELECT *
FROM person`
```
There's no real intent to preserve a leading space.

The main purpose in keeping the white-space was to help with determining that ampersand-prefixed identifiers were from our DSL, and lot logical "AND", however this turns out to be moot, so they are no longer returned.

manadart commented on parse/lexer.go:
```diff
@@ -0,0 +1,213 @@
+		return tok
+	}
+
+	l.nextChar()
```
Indeed.

manadart commented on parse/lexer.go:
```diff
@@ -0,0 +1,213 @@
+		l.nextChar()
+	}
+
+	return string(l.input[pos:l.offset])
```
Ended up doing this.

manadart commented on parse/lexer.go:
```diff
@@ -0,0 +1,213 @@
+				next := l.peek()
+				if next == ' ' || next == '\n' || next == 0 || next == ';' || next == ',' || next == ')' {
+					l.nextChar()
+					return string(ret)
```
In the initial implementation, the parser will ignore these. They'll go though verbatim. We may entertain positional substitution in future, but it is eschewed for now.

manadart commented on parse/lexer.go:
```diff
@@ -0,0 +1,213 @@
+
+	for isDigit(l.char) || l.char == '.' {
+		if l.char == '.' {
+			if l.peek() == '.' {
```
I've re-implemented this. We're only allowing one decimal.

manadart commented on parse/lexer.go:
```diff
@@ -0,0 +1,213 @@
+				// TODO (manadart 2022-06-08): This is very naive and will not
+				// correctly handle some edge cases. Review for robustness.
+				next := l.peek()
+				if next == ' ' || next == '\n' || next == 0 || next == ';' || next == ',' || next == ')' {
```
I've reworked it based on pairing up quotes.

manadart commented:
@benhoyt Thanks for the feedback. I've implemented some of your suggestions.

Note that the code in this repository outside of the this patch no longer represents the DSL that it will implement. What we're parsing is specified [here](https://docs.google.com/document/d/1FbRZisjK0z7shy8m1cHzzEMDpxGwsK-54xDpf9HY4Xo). So we're no longer just subbing out braced contents.

I looked at a couple of the examples you posted. We don't actually need a full SQL parser, which would be overkill. Most of the tokens we detect will be left untouched. What we're doing is detecting type references of inputs and outputs, then expanding/replacing based on reflection info.

Unless you feel super-strongly about anything else, I'll probably merge this and begin to iterate.

benhoyt commented:
@manadart Thanks -- good fixes. I don't feel super-strongly about anything else. However, the rune parsing wasn't quite what I had in mind, and currently breaks on non-ASCII, due to using `DecodeLastRuneInString` on a single byte. What you want is `utf8.DecodeRuneInString(l.input[l.readOffset:])` (the string slice is essentially free).

Here's a test that shows this bug:

```go
func TestLexerUnicode(t *testing.T) {
	stmt := "Persön.*"
	expected := []string{"Persön", ".", "*"}
	assert.Equal(t, expected, stringsFromTokens(tokensForStatement(stmt)))
}
```

And what I think are the fixes to `nextChar` and `peek`:

```go
// nextChar reads the next character from the
// input and increments the read offset.
func (l *Lexer) nextChar() {
	if l.readOffset >= len(l.input) {
		l.char = 0
		return
	}

	var size int
	l.char, size = utf8.DecodeRuneInString(l.input[l.readOffset:])
	if l.char == '\n' {
		l.line++
		l.column = 0
	}
	l.column++
	l.offset = l.readOffset
	l.readOffset += size
}

// peek returns the next rune to be read without moving the lexer forward.
func (l *Lexer) peek() rune {
	if l.readOffset >= len(l.input) {
		return 0
	}

	peek, _ := utf8.DecodeRuneInString(l.input[l.readOffset:])
	return peek
}
```

benhoyt commented:
Oh, I do feel strongly about `char == 39`. Magic numbers! :-) Besides, the body of `isSingleQuote` is shorter than its name, so we might as well just say `l.char == '\''` inline. Or at least get rid of the magic number 39.

fernape committed:
Add Parser skeleton.

Also, add definition for the `Part` interface to be implemented by the
parts that will make the AST.

The parser contains two indexes to keep track of the current position
and the last parsed position.

Aflynn50 commented on internal/parser/parser.go:
```diff
@@ -0,0 +1,47 @@
+
+// Parser defines the SDL parser
+type Parser struct {
+	parts   []Part // Array of parts. Our simple AST
```
The name 'parts' could be clearer, perhaps something like queryPart since it will often appear without it being super obvious its part of the Parser

fernape replied:
QueryPart is clearer, thanks.

Aflynn50 commented on internal/parser/parser.go:
```diff
@@ -0,0 +1,47 @@
+	parts   []Part // Array of parts. Our simple AST
+	input   string // statement to be parsed
+	parsed  int    // position of the last parsed chunk
+	skipped int    // current position of the parser
```
Could this not be 'currentPos' instead of skipped since its the current position of the parser?

fernape replied:
This one is a bit more difficult to explain. We can of course change `parsed` and `skipped` to something else, but we should choose something that truly reflects the intention. `parsed`  marks the position up to which we have recognized the input stream and generated `QueryParts` for. `skipped` on the other hand represents the temporary cursor we use to [advance through the input and see if we recognize something else](https://github.com/fernape/sqlairexperiment/blob/main/README.md#parsing).

fernape replied:
Changed this to `pos` as is very idiomatic and we keep it short.

Aflynn50 commented on internal/parser/parser.go:
```diff
@@ -0,0 +1,47 @@
+	return nil, nil
+}
+
+// Part defines a simple interface for all the different parts that
```
Again this could be renamed to QueryPart or something similar to make it a little more specific

fernape replied:
Changed!

Aflynn50 commented on internal/parser/parser.go:
```diff
@@ -0,0 +1,47 @@
+//
+// would be represented as:
+//
+// [stringPart outputPart stringPart inputPart]
```
These implementations of the Part interface don't exist yet, but I guess it's still pretty clear

fernape replied:
Yes. And names will probably change. `stringPart` might lead to confusion as in reality it has nothing to do with strings. It is more like a `verbatimPart`, `passthroughPart`, etc.

Aflynn50 commented on internal/parser/parser.go:
```diff
@@ -0,0 +1,47 @@
+//
+// [stringPart outputPart stringPart inputPart]
+type ParsedExpr struct {
+	parts []Part
```
It's a bit confusing having a field with the same name as the Parser field 'parts'. This could be renamed to parsedQueryParts maybe, since that is what it contains.

fernape replied:
I've removed the attribute from the parser. It was there in the experiment just to keep track of the parts and then it was copied to the returned `ParsedExpression`. It definitely does not belong in the `Parser`.

Aflynn50 commented on internal/parser/parser.go:
```diff
@@ -0,0 +1,47 @@
+type Parser struct {
+	parts   []Part // Array of parts. Our simple AST
+	input   string // statement to be parsed
+	parsed  int    // position of the last parsed chunk
```
This name could also be expanded to be more descriptive

fernape replied:
`lastParsedPos` should be a better name.

Aflynn50 reviewed (commented):
A lot of the design decisions made here are dependent on the future implementation of the parser and AST, they should make more sense once those have been added as well (e.g. 'parts' being a slice instead of a tree).

manadart commented on internal/parser/parser.go:
```diff
@@ -0,0 +1,47 @@
+package parser
```
Nit: I like "parse" for the package. There's a tendency towards using verb-er for everything.

fernape replied:
Done, thanks!

manadart commented on internal/parser/parser.go:
```diff
@@ -0,0 +1,47 @@
+
+// Part defines a simple interface for all the different parts that
+// compose a ParsedExpr
+type Part interface {
```
I'd declare this at the top ahead of it's first usage in `Parser`

fernape replied:
Done!

manadart reviewed (approved):
Good start.

We need to set up some workflows to run tests for us automatically and gate landing, maybe a merge bot too.

Let's institute a policy now of using merge commits, not squash/rebase. Before merging this one, please squash it though.
